realm/realm-library/src/main/cpp/io_realm_internal_OsObject.cpp

```diff
@@ -152,14 +155,98 @@ static void finalize_object(jlong ptr)
     delete reinterpret_cast<ObjectWrapper*>(ptr);
 }
 
+static inline size_t do_create_row(jlong shared_realm_ptr, jlong table_ptr)
+{
+    auto& shared_realm = *(reinterpret_cast<SharedRealm*>(shared_realm_ptr));
+    auto& table = *(reinterpret_cast<realm::Table*>(table_ptr));
+    shared_realm->verify_in_write();
+    return table.add_empty_row();
+}
+
+template <class T>
+static void throw_duplicated_primary_key_exception(JNIEnv* env, T value)
+{
+    static JavaClass dup_pk_exception(env, "io/realm/exceptions/RealmPrimaryKeyConstraintException");
+    env->ThrowNew(dup_pk_exception, format("Primary key value already exists: %1 .", value).c_str());
+}
+
+static inline size_t do_create_row_with_primary_key(JNIEnv* env, jlong shared_realm_ptr, jlong table_ptr,
+                                                    jlong pk_column_ndx, jlong pk_value, jboolean is_pk_null)
+{
+    auto& shared_realm = *(reinterpret_cast<SharedRealm*>(shared_realm_ptr));
+    auto& table = *(reinterpret_cast<realm::Table*>(table_ptr));
+    shared_realm->verify_in_write(); // throws
+    if (is_pk_null && !TBL_AND_COL_NULLABLE(env, &table, pk_column_ndx)) {
+        return npos;
+    }
+
+    if (is_pk_null) {
+        if (table.find_first_null(pk_column_ndx) != npos) {
+            throw_duplicated_primary_key_exception(env, "'null'");
+            return npos;
+        }
+    }
+    else {
+        if (table.find_first_int(pk_column_ndx, pk_value) != npos) {
+            throw_duplicated_primary_key_exception(env, reinterpret_cast<long long>(pk_value));
+            return npos;
+        }
+    }
+
+    size_t row_ndx = table.add_empty_row();
+
+    if (is_pk_null) {
+        table.set_null_unique(pk_column_ndx, row_ndx);
+    }
+    else {
+        table.set_int_unique(pk_column_ndx, row_ndx, pk_value);
+    }
+    return row_ndx;
+}
+
+static inline size_t do_create_row_with_primary_key(JNIEnv* env, jlong shared_realm_ptr, jlong table_ptr,
+                                                    jlong pk_column_ndx, jstring pk_value)
+{
+    auto& shared_realm = *(reinterpret_cast<SharedRealm*>(shared_realm_ptr));
+    auto& table = *(reinterpret_cast<realm::Table*>(table_ptr));
+    JStringAccessor str_accessor(env, pk_value); // throws
+    shared_realm->verify_in_write();             // throws
```

Comment:
The only way for users to distinguish `IllegalStateException`s is to check the message in it.
I guess some users in the world is actually doing that. Since the message itself is not a public interface, I don't think that changing it is a breaking change, but shouldn't we notice it at least?